packages/block-editor/src/store/actions.js

```diff
@@ -987,3 +987,21 @@ export function* insertAfterBlock( clientId ) {
 	);
 	yield insertDefaultBlock( {}, rootClientId, firstSelectedIndex + 1 );
 }
+
+/**
+ * Generator that toggles the highlighted block state.
+ *
+ * @param {string} clientId The block's clientId.
+ * @param {boolean} isHighlighted The highlight state.
+ */
+export function toggleBlockHighlight( clientId, isHighlighted ) {
+	if ( isHighlighted ) {
+		return {
+			type: 'HIGHLIGHT_BLOCK',
+			clientId,
+		};
+	}
+	return {
+		type: 'UNHIGHLIGHT_BLOCK',
+	};
```

Comment:
Two thoughts:

- Do we need separate action type for this, if we're already expressing the function as a single "toggle" operation? Typically I would expect the action creator to align to its own type, i.e. `TOGGLE_BLOCK_HIGHLIGHT`, or vice-versa, if we have separate types, then separate functions `highlightBlock`, `unhighlightBlock`. I have a strong preference toward the former (a single action type).
- Related to this point, I've been wondering if there's a possibility of race conditions where a block's unmounting (and subsequent "unhighlight" action) might occur only _after_ another block has opted to become "highlighted", and thus wrongly unhighlight the next highlighted block. Having a single action type might help with this, because if we have:

```js
return {
	type: 'TOGGLE_BLOCK_HIGHLIGHT',
	clientId,
	isHighlighted,
};
```

...then the reducer could verify that the "unhighlight" intent matches its own expectations for what the highlighted block is:

```js
function highlightedBlock( state = null, action ) {
	if ( action.type === 'TOGGLE_BLOCK_HIGHLIGHT' ) {
		if ( action.isHighlighted ) {
			return action.clientId;
		} else if ( state === action.clientId ) {
			return null;
		}
	}

	return state;
}
```

(My example could perhaps be expressed more elegantly)